pkg/trace/sampler/exception_sampler.go

```diff
@@ -0,0 +1,207 @@
+package sampler
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/traceutil"
+	"golang.org/x/time/rate"
+)
+
+const (
+	// cardinalityLimit limits the number of spans considered per combination of (env, service).
+	cardinalityLimit = 1000
+	// defaultTTL limits the frequency at which we sample a same span (env, service, name, rsc, ...).
+	defaultTTL = 2 * time.Minute
+	// priorityTTL allows to blacklist p1 spans that are sampled entirely, for this period.
+	priorityTTL = 10 * time.Minute
+	// ttlRenewalPeriod specifies the frequency at which we will upload cached entries.
+	ttlRenewalPeriod = 1 * time.Minute
+	// exceptionSamplerTPS traces per second allowed by the rate limiter.
+	exceptionSamplerTPS = 5
+	// exceptionSamplerBurst sizes the token store used by the rate limiter.
+	exceptionSamplerBurst = 50
+	exceptionKey          = "_dd.exception"
+)
+
+// ExceptionSampler samples traces that are not caught by the Priority sampler.
+// It ensures that we sample traces for each combination of
+// (env, service, name, resource, error type, http status) seen on a top level or measured span
+// for which we did not see any span with a priority > 0 (sampled by Priority).
+// The resulting sampled traces will likely be incomplete.
```

Comment:
Nit: Worth adding as a comment here that these are flagged with `exceptionKey` as a span metric.